File: csharp/2015/day05b/Program.cs
using System;
using System.IO;

class Program
{
    static void Usage()
    {
        Console.Error.WriteLine("usage: dotnet run <input file>");
        Environment.Exit(1);
    }

    static bool Prop1(string str)
    {
        for ( int i = 0; i < str.Length - 3; i++ )
        {
            string firstPair = str.Substring(i, 2);
            for ( int j = i + 2; j < str.Length - 1; j++ )
            {
                string secondPair = str.Substring(j, 2);
                if ( firstPair.Equals(secondPair) )
                {
                    return true;
                }
            }
        }
        return false;
    }

    static bool Prop2(string str)
    {
        for ( int i = 0; i < str.Length - 2; i++ )
        {
            if ( str[i] == str[i + 2] )
            {
                return true;
            }
        }
        return false;
    }

    static int Process(string filename)
    {
        int count = 0;

        try
        {
            using ( StreamReader reader = new StreamReader(filename) )
            {
                string? line;
                while ( (line = reader.ReadLine()) != null )
                {
                    if ( Prop1(line) && Prop2(line) )
                    {
                        count++;
                    }
                }
            }
        }
        catch ( IOException e )
        {
            Console.Error.WriteLine("Error reading file: " + e.Message);
            Environment.Exit(1);
        }

        return count;
    }

    static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Usage();
        }

        string filename = args[0];
        int result = Process(filename);
        Console.WriteLine("result = " + result);
    }
}

File: fsharp/2015/day05b/Program.fs
open System
open System.IO

let usage() =
    Console.Error.WriteLine("usage: dotnet run <input file>")
    Environment.Exit(1)

let prop1(str: string) =
    let rec f1 s =
        let rec f2 a b s =
            match s with
            | c :: d :: rest -> if (a = c) && (b = d) then
                                    true
                                else
                                    f2 a b (d :: rest)
            | _ -> false
        match s with
        | a :: b :: rest -> if f2 a b rest then
                                true
                            else f1 (b :: rest)
        | _ -> false
    str |> Seq.toList |> f1

let prop2(str: string) =
    let rec f s =
        match s with
        | a :: b :: c ::  _ when a = c -> true
        | a :: b :: c :: rest -> f (b :: c :: rest)
        | _ -> false
    str |> Seq.toList |> f

let processFile(filename: string) =
    let f acc (line: string) =
        if (prop1 line) && (prop2 line) then
            acc + 1
        else
            acc

    try
        let content = File.ReadAllText(filename)
        content.Split([|'\n'|], StringSplitOptions.RemoveEmptyEntries)
        |> Seq.fold f 0
    with
    | :? IOException as e ->
        Console.Error.WriteLine("Error reading file: " + e.Message)
        Environment.Exit(1)
        0

[<EntryPoint>]
let main args =
    if args.Length < 1 then
        usage()

    let filename = args.[0]
    let result = processFile filename
    Console.WriteLine("result = " + result.ToString());
    0 
